### PR TITLE
feat: update gap fill planner rule to use LOCF

### LINKED ISSUE #7071: Add `LOCF` scalar function and update `HandleGapFill` optimizer rule for use in SQL
Part of #6594.

`LOCF` stands for "last observation carried forward" and is used to indicate that a column should have it's gaps filled with a previous value in the column. The doc linked from #6594 has more info.

The work of this PR is to update [this logical optimizer rule](https://github.com/influxdata/influxdb_iox/blob/main/iox_query/src/logical_optimizer/handle_gapfill.rs) to:
- look for a `Projection` that appears over a `GapFill` node that was just created
- find any calls to `LOCF` in that projection
- if any are found, "push" them into the `GapFill` node by updating its parameters to reflect [`FillStrategy::Prev`](https://github.com/influxdata/influxdb_iox/blob/af71e3295caffd3f9861c74b2c394ee47411456b/iox_query/src/exec/gapfill/mod.rs#L71) (or 'PrevNullAsMissing`) for the appropriate columns

### PR DESCRIPTION
**Pull request by wolffcm**
Closes #7071.

This PR:
- Updates the planner rule for gap filling to look for `LOCF`[^1] and update the fill strategy as needed
- Modifies how plans are shown for `EXPLAIN` to reflect when `LOCF` is pushed down
- Adds end-to-end tests for gap filling with previous values (which is now actually possible from SQL)

[^1]: **L**ast **O**bservation **C**arried **F**orward

### CODE REVIEW

File: influxdb_iox/tests/query_tests2/cases/in/gapfill.sql [hunk 1]
```diff
@@ -48,3 +48,21 @@ from cpu
 where time between timestamp '2000-05-05T12:00:00Z' and timestamp '2000-05-05T12:59:00Z'
 group by minute, region;
 
+-- gap filling with previous value using LOCF
+-- IOX_COMPARE: uuid
+EXPLAIN SELECT
+  region,
+  date_bin_gapfill(interval '10 minute', time, timestamp '1970-01-01T00:00:00Z') as minute,
+  locf(avg(cpu.user))
```
Comment (stuartcarnie) [inline influxdb_iox/tests/query_tests2/cases/in/gapfill.sql:56]: 💯

File: iox_query/src/logical_optimizer/handle_gapfill.rs [hunk 2]
```diff
@@ -9,41 +9,61 @@ use datafusion::{
...
+/// - Convert `DATE_BIN_GAPFILL()` to `DATE_BIN()`
+/// - Create a `GapFill` node that fills in gaps in the query
+/// - The range for gap filling is found by analyzing any preceding `Filter` nodes
+///
+/// If there is a `Projection` above the `GapFill` node that gets created:
+/// - Look for calls to gap-filling functions like `LOCF`
+/// - Push down these functions into the `GapFill` node, updating the fill strategy for the column.
+///
+/// Note: both `DATE_BIN_GAPFILL` and `LOCF` are functions that don't have implementations.
+/// This rule must rewrite the plan to get rid of them.
```
Comment (stuartcarnie) [inline iox_query/src/logical_optimizer/handle_gapfill.rs:66]: docs are 💯

File: iox_query/src/logical_optimizer/handle_gapfill.rs [hunk 3]
```diff
@@ -328,29 +349,97 @@ impl TreeNodeRewriter for DateBinGapfillRewriter {
...
+        // then there is nothing to do.
+        return Ok(None)
+    };
+
+    let fill_cols: Vec<(&Expr, FillStrategy)> = proj_exprs
+        .iter()
+        .filter_map(|e| match e {
+            Expr::ScalarUDF { fun, args } if fun.name == LOCF_UDF_NAME => {
+                let col = &args[0];
+                Some((col, FillStrategy::Prev))
```
Comment (stuartcarnie) [inline iox_query/src/logical_optimizer/handle_gapfill.rs:373]: Do we plan to support an option to carry only non-null values forward, like InfluxQL?

**Review by stuartcarnie — APPROVED** [review]
Looks great, @wolffcm!

File: iox_query/src/logical_optimizer/handle_gapfill.rs (hunk 3 above)
Comment (wolffcm) [reply iox_query/src/logical_optimizer/handle_gapfill.rs:373]: Definitely, I need to file an issue for it.

Comment (wolffcm) [reply iox_query/src/logical_optimizer/handle_gapfill.rs:373]: I filed https://github.com/influxdata/influxdb_iox/issues/7368 for this

**Pull request merged**